src/read_write/node_iterator.rs

```diff
@@ -48,6 +52,67 @@ impl NodeIterator {
             batch_size,
         }
     }
+
+    pub fn from_data_provider<Id: ToString>(
+        data_provider: &dyn DataProvider,
+        attributes: &[&str],
+        encoding: Encoding,
+        id: &Id,
+        num_points: usize,
+        batch_size: usize,
+    ) -> Result<Self> {
+        if num_points == 0 {
+            return Ok(NodeIterator::default());
+        }
+
+        let attribute_data_types = match data_provider.meta_proto() {
+            Ok(ref meta) if meta.has_s2() => meta
+                .get_s2()
+                .get_attributes()
+                .iter()
+                .map(|a| {
+                    AttributeDataType::from_proto(a.data_type)
+                        .map(|data_type| (a.name.clone(), data_type))
+                })
+                .collect::<Result<HashMap<String, AttributeDataType>>>()?,
+            _ => vec![
+                ("color".to_string(), AttributeDataType::U8Vec3),
+                ("intensity".to_string(), AttributeDataType::F32),
+            ]
+            .into_iter()
+            .collect(),
+        };
+
+        let mut all_reads =
+            data_provider.data(&id.to_string(), &[&["position"], attributes].concat())?;
+        let position_read = all_reads
+            .remove("position")
+            .ok_or_else(|| -> Error { "No position reader available.".into() })?;
+
+        let mut attribute_readers = HashMap::new();
+        for attribute in attributes {
+            match all_reads.remove(*attribute) {
+                Some(data) => {
```

Comment:
Wouldn't the `data_provider.data()` call have failed already if this wasn't Some? If so, how about `unwrap()` with a comment?

Reply:
Done.